rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/NodesRecoveryManager.java

```diff
@@ -75,8 +99,202 @@ public NodesLockRestorationManager apply(RMCore rmCore) {
+            } else {
+                recoverNodes = true;
+                nodeSourceToDeploy.setInfrastructureManager(im);
+            }
+        }
+        return recoverNodes;
+    }
+
+    protected void recoverNodes(NodeSource nodeSource) {
+        logger.info(START_TO_RECOVER_NODES); // this log line is important for performance tests
+        int lookUpTimeout = PAResourceManagerProperties.RM_NODELOOKUP_TIMEOUT.getValueAsInt();
+        String nodeSourceName = nodeSource.getName();
+        this.logWarnIfNodeSourceHasNoNode(nodeSource, nodeSourceName);
+
+        Collection<RMNodeData> nodesData = this.rmCore.getDbManager().getNodesByNodeSource(nodeSourceName);
+        logger.info("Number of nodes found in database for node source " + nodeSourceName + ": " + nodesData.size());
+
+        List<RMNode> recoveredEligibleNodes = Collections.synchronizedList(new ArrayList<RMNode>());
+        Map<NodeState, Integer> recoveredNodeStatesCounter = new HashMap<>();
+
+        // for each node found in database, try to lookup node or recover it
+        // as down node
+        for (RMNodeData rmNodeData : nodesData) {
+            String nodeUrl = rmNodeData.getNodeUrl();
+
+            Node node = this.tryToLookupNode(nodeSource, lookUpTimeout, nodeUrl);
+            RMNode rmnode = this.recoverRMNode(nodeSource, recoveredNodeStatesCounter, rmNodeData, nodeUrl, node);
+
+            if (this.isEligible(rmnode)) {
+                recoveredEligibleNodes.add(rmnode);
+            }
+        }
+
+        this.rmCore.recoverEligibleNodes(recoveredEligibleNodes);
+
+        this.logNodeRecoverySummary(nodeSourceName, recoveredNodeStatesCounter, recoveredEligibleNodes.size());
+    }
+
+    private boolean existPersistedNodes(String nodeSourceName) {
+        boolean existPersistedNodes = false;
+
+        Collection<RMNodeData> nodesData = this.rmCore.getDbManager().getNodesByNodeSource(nodeSourceName);
+        if (nodesData.isEmpty()) {
+            logger.info("No node found in database for node source: " + nodeSourceName);
+        } else {
+            existPersistedNodes = true;
+        }
+
+        return existPersistedNodes;
+    }
+
+    private void recoverNodeSourceSuccessfullyOrRemove(NodeSourceData nodeSourceData, String nodeSourceName) {
+        try {
+            logger.info("Recover node source " + nodeSourceName);
+            // retrieve node source status
+            boolean deployNodeSource = false;
+            if (nodeSourceData.getStatus().equals(NodeSourceStatus.NODES_DEPLOYED)) {
+                // reset node source status to be able to deploy again
+                nodeSourceData.setStatus(NodeSourceStatus.NODES_UNDEPLOYED);
+                deployNodeSource = true;
+            }
+            this.rmCore.prepareNodeSource(nodeSourceData);
+            if (deployNodeSource) {
+                this.rmCore.deployNodeSource(nodeSourceName);
+            }
+        } catch (Throwable t) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Catch Exception instead of Throwable. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1181)